scripts/Auth_scripts/auth.ps1

```diff
@@ -0,0 +1,36 @@
+Function authUser($url)
+{
+    while($true)
+    {
+        $username = Read-Host 'Please enter your username'
+        $password = Read-Host 'Please enter your password' -AsSecureString
+        $body = @{
+            Username=$username
+            Password=[Runtime.InteropServices.Marshal]::PtrToStringAuto([Runtime.InteropServices.Marshal]::SecureStringToBSTR($password))
+        } | ConvertTo-Json
+        $headers = @{
+            'Content-Type' = 'application/json'
+        }
+        try
+        {
+            $response = Invoke-WebRequest -Uri $url/api/account/login -Method POST -Body $body -Headers $headers
+        }
+        catch
+        {
+            if (  $_.Exception.response.StatusCode -eq "Unauthorized" )
+            {
+                Write-Warning 'Incorrect username or password'
+                continue
+            }
+            else 
+            {
+               throw
+            }
+        }
+        break
+    }
+    $token = (($response.content | ConvertFrom-Json).token).ToString()
+    $env:TOKEN = $token

```

Comment:
Do not store any not operating system relevant variables in the environment variables! Instead, you can use global variables in your scripts mostly. Here is an example using cmdlet: Set-Variable -Name "accessToken" -Scope global -Value $accessToken. 
Please, rename the token variable to the accessToken, so it will be more clear.

Reply:
Actually the global variables is also not the most reliable things to have, because they provide unnecessary dependencies and one day they can break the user environment. The better approach to this is to let your function returns actually the values you are looking for (accessToken). 

someFunction {
   $x = 0
   return $x
}

$y = someFunction
In case you have to return few variables, you always can use object as return type.

Reply:
Changed

Reply:
The problem with return statement is that every time you want to to anything with script you will have to authorize again and again. To avoid such behavior we actually need to store it somewhere and for that I choose environment variables, but global sounds even better.

Reply:
Can't you store as local variable the function result in your main.ps1 and then if needed provide it to the next function as parameter?

Reply:
You can, but what if you need to start the script all over again the minute later? You'll then have to authorize again. And that's every time you run main.ps1. Or do we only want save the token for current session. I mean Is it okay that user has to authorize every time he starts main.ps1?

Reply:
What exactly do you mean by user session here? Each time you close PS and open it again? In this case - yes, you should input the credentials each time you close and open power shell again. Have you tried to do non single session (when you close and open PS) with global variables (not the environment variables). I presume you will not be able to do that.

Reply:
By session I mean calling main.ps1 once, every time you call main.ps1 you'll have to input login and password to get token. We can avoid that by using environment or global variables, local variables will disappear after executing the srcipt once. And environment variables also will disappear if you close PS - FYI, actually I believe any variables.

Reply:
There is no point to execute main script few times actually. The flow would be something like this main -> (choose what you want to do) -> fucntion -> (complete some job) - in case you want to proceed one more action it is okay to login one more time (anyway we will change the token expiration time something to 5-10 mins).

Reply:
Okay for example you ran main.ps1, got the pet data also got the token along the way. Three minutes later you run the same main.ps1 from the beginning. The Question: Do we want to have the token already (that was saved from previous interaction with main.ps1) or user should authorize again?

Reply:
Yea, okay. Lets proceed with global variable here.